rfcs/widget-restructure-rfc-1.md

```diff
@@ -0,0 +1,667 @@
+    let id =  self.current_id;
+    let (should_rerender, child_id) = self.context
+      .widget_manager
+      .create_widget(widget_index, widget.clone(), Some(id));
+    self.widget_tree.add(child_id, Some(id));
+    if should_rerender {
+        let mut child_widget = self.context.widget_manager.take(child_id);
+        let mut context = KayakContextRef { context: self.context, current_id: id };
+        child_widget.render(context);
+        self.context.widget_manager.repossess(child_widget);
+    }
+  }
+
+  pub fn commit(mut self) {
+    let tree = self.widget_tree.take();
+    let id =  self.current_id;
+    let changes = self.context.widget_manager.tree.diff_children(&tree, id);
+    self.context.widget_manager.tree.merge(&tree, id, changes);
+  }
+}
+```
+
+> This was added in Version 2 of this RFC, but Version 3 internalizes a bit more of the logic, specifically by storing the `WidgetTree` in context.
+
+And results in usage like so:
+
+```rust
+// fragment.rs
+
+fn render(&mut self, mut context: KayakContextRef) {
+  if let Some(children) = self.children.take() {
+    context.add_children(children);
+  }
+  
+  context.commit();
+}
+
+// vec.rs
+
+fn render(&mut self, mut context: KayakContextRef) {
+  for (index, item) in self.data.iter().enumerate() {
+      context.add_widget(item, index);
+  }
+  
+  context.commit();
+}
+```
+
+<br />
+
+<br />
+
+#### Functional Widgets
+
+The other form of widget we need to consider are functional widgets. These are the widgets to use when customization isn't a huge concern and you just need to define a working widget, and this is probably what most people will use when defining their own widgets.
+
+In order to allow custom widgets to properly integrate with functional widgets and vice-versa, there are a few changes that need to be made.
+
+Consider the following widget:
+
+```rust
+#[widget]
+fn MyWidget(foo: Option<i32>, bar: String) {
+  // ...
+}
+```
+
+Currently this generates the following code:
+
+```rust
+#[derive(Derivative)]
+#[derivative(Default, Debug, PartialEq, Clone)]
+struct MyWidget {
+	pub id: kayak_core::Index,
+  pub foo: Option<i32>,
+  pub bar: String,
+  #[derivative(Default(value = "None"))]
+	pub styles: Option<Style>,
+  #[derivative(Debug = "ignore", PartialEq = "ignore")] 
+  pub children: Children,
+	#[derivative(Default(value = "None"), Debug = "ignore", PartialEq = "ignore")] 
+  pub on_event: Option<kayak_core::OnEvent>,
+}
+```
+
+However, we need this:
+
+```rust
+#[derive(Derivative)]
+#[derivative(Default, Debug, PartialEq, Clone)]
+struct MyWidget {
+  pub id: kayak_core::Index,
+  #[derivative(Debug = "ignore", PartialEq = "ignore")] 
+  pub children: Children,
+  pub props: MyWidgetProps,
+}
+```
+
+In order to properly get the props struct, we can instead define our widget like so:
+
+```rust
+#[widget]
+fn MyWidget(props: MyWidgetProps) {
+  let WidgetProps {foo, bar} = props;
```

Comment:
Internally we should probably `Clone` props so we have values at this point. I think this code would resolve to:
```rust
fn render(...) {
  let props = self.props.clone();
  let WidgetProps { foo, bar } = props;
  
  ...
}
```